docs/tech-notes/range-merges.md

```diff
@@ -0,0 +1,259 @@
+_every_ replica of the RHS, indicating that no traffic is possibly being served
+on the RHS, does the coordinator commit the merge transaction.
+
+Like with splits, the merge transaction is committed with a special "commit
+trigger" that instructs the receiving store to update its in-memory bookkeping
+to match the updates to the range descriptors in the transaction. The moment the
+merge transaction is considered committed, the merge is complete!
+
+The remaining sections cover each of these steps in more detail.
+
+## Preconditions
+
+Not any two ranges can be merged. The first and most obvious criterion is that
+the two ranges must be adjacent. Suppose a simplified cluster that has only
+three ranges, A, B, and C:
+
+```
++-----------+-----+
+|  A  |  B  |  C  |
++-----------+-----+
+```
+
+Ranges A and B can be merged, as can ranges B and C, but not ranges A and C, as
+they are not adjacent. Note that adjacent ranges are equivalently referred to
+as "neighbors", as in, range B is range A's right-hand neighbor.
+
+The second criterion is that the replica sets must be aligned. To illustrate,
+consider a four node cluster with the default 3x replication. The allocator has
+attempted to balance ranges as evenly as possible:
+
+```
+Node  1    Node  2     Node  3     Node  4
++-----+    +-----+     +-----+     +-----+
+| P Q |    |  P  |     |  Q  |     | P Q |
++-----+    +-----+     +-----+     +-----+
+```
+
+Notice how node 2 does not have a copy of Q, and node 3 does not have a copy of
+P. These replica sets are considered "misaligned." Aligning them requires
+rebalancing Q from node 3 to node 2, or rebalancing P from node 2 to node 3:
+
+```
+Node  1    Node  2     Node  3     Node  4
++-----+    +-----+     +-----+     +-----+
+| P Q |    | P Q |     |     |     | P Q |
++-----+    +-----+     +-----+     +-----+
+
+Node  1    Node  2     Node  3     Node  4
++-----+    +-----+     +-----+     +-----+
+| P Q |    |     |     | P Q |     | P Q |
++-----+    +-----+     +-----+     +-----+
+```
+
+## Initiating a merge
+
+A merge is initiated by sending a AdminMerge request to a range. Like other
+admin commands, DistSender will automatically route the request to the
+leaseholder of the range, but there is no guarantee that the store will retain
+its lease while the admin command is executing.
+
+Note that an AdminMerge request takes no arguments, as there is no choice in
+what range will be merged. The recipient of the AdminMerge will always be the
+LHS, subsuming range, and its right neighbor at the moment that the AdminMerge
+command begins executing will always be the RHS, subsumed range.
+
+It would perhaps have been wise to require that the AdminMerge specify the
+descriptor of the intented subsumption target. It is, at present, possible for
+someone to send an admin merge to P, intending to merge P's right neighbor Q,
+but Q can split in the meantime and so instead P merges with a different range.
```

Comment:
I assume it's also possible that a different client had already asked P and Q to merge, and so a second/repeated request could cause PQ and R to merge?